samples/quickstart_oauth2.js

```diff
@@ -0,0 +1,195 @@
+      // Create an OAuth client to authorize the API call.
+      const oAuth2Client = new OAuth2Client(
+        keys.web.client_id,
+        keys.web.client_secret,
+        // A hardcoded callback URL used in this sample app.
+        'http://localhost:3000/oauth2callback'
+      );
+
+      // Generate the url that will be used for the consent dialog.
+      const authorizeUrl = oAuth2Client.generateAuthUrl({
+        access_type: 'offline',
+        scope: SCOPES.join(' '),
+      });
+
+      // Open an http server to accept the oauth callback. In this example, the
+      // only request to our webserver is to /oauth2callback?code=<code>
+      const server = http
+        .createServer(async (req, res) => {
+          try {
+            if (req.url.indexOf('/oauth2callback') > -1) {
+              // Acquire the code from the querystring, and close the web
+              // server.
+              const qs = new url.URL(req.url, 'http://localhost:3000')
+                .searchParams;
+              const code = qs.get('code');
+              console.log(`Code is ${code}`);
+              res.end(
+                'Authentication successful! Please return to the console.'
+              );
+              server.destroy();
+
+              // Now that we have the code, use that to acquire tokens.
+              const r = await oAuth2Client.getToken(code);
+              // Make sure to set the credentials on the OAuth2 client.
+              oAuth2Client.setCredentials(r.tokens);
+              console.info('Tokens acquired.');
+              resolve(oAuth2Client);
+            }
+          } catch (e) {
+            reject(e);
+          }
+        })
+        .listen(3000, () => {
+          // Open the browser to the authorize url to start the workflow.
+          // This line will not work if you are running the code in the
+          // environment where a browser is not available. In this case,
+          // copy the URL and open it manually in a browser.
+          console.info(`Opening the browser with URL: ${authorizeUrl}`);
+          open(authorizeUrl, {wait: false}).then(cp => cp.unref());
+        });
+      destroyer(server);
+    });
+  }
+
+  // Runs a simple report using the supplied Data API client instance.
+  async function runReport(analyticsDataClient) {
+    const [response] = await analyticsDataClient.runReport({
+      entity: {
+        propertyId: propertyId,
+      },
+      dateRanges: [
+        {
+          startDate: '2020-03-31',
+          endDate: 'today',
+        },
+      ],
+      dimensions: [
+        {
+          name: 'city',
+        },
+      ],
+      metrics: [
+        {
+          name: 'activeUsers',
+        },
+      ],
+    });
+
+    console.log('Report result:');
+    response.rows.forEach(row => {
+      console.log(row.dimensionValues[0], row.metricValues[0]);
+    });
+  }
+
+  // Starts the OAuth2 flow and runs the report.
+  getOAuth2Client().then(oAuth2Client => {
```

Comment:
Is it possible to rewrite this using `await`? Generally it's better not to mix `then` and `await(s)`

Reply:
I believe this is fixed now.